src/java.desktop/share/classes/sun/font/HBShaper.java

```diff
@@ -0,0 +1,663 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+package sun.font;
+
+import java.awt.geom.Point2D;
+import sun.font.GlyphLayout.GVData;
+import sun.java2d.Disposer;
+import sun.java2d.DisposerRecord;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import static java.lang.foreign.MemorySegment.NULL;
+import java.lang.foreign.SequenceLayout;
+import java.lang.foreign.StructLayout;
+import java.lang.foreign.SymbolLookup;
+import java.lang.foreign.UnionLayout;
+import static java.lang.foreign.ValueLayout.*;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.VarHandle;
+
+import java.util.Optional;
+import java.util.WeakHashMap;
+
+public class HBShaper {
+
+    /*
+     * union _hb_var_int_t {
+     *     uint32_t u32;
+     *     int32_t i32;
+     *     uint16_t u16[2];
+     *     int16_t i16[2];
+     *     uint8_t u8[4];
+     *     int8_t i8[4];
+     * };
+     */
+   private static final UnionLayout VarIntLayout = MemoryLayout.unionLayout(
+        JAVA_INT.withName("u32"),
+        JAVA_INT.withName("i32"),
+        MemoryLayout.sequenceLayout(2, JAVA_SHORT).withName("u16"),
+        MemoryLayout.sequenceLayout(2, JAVA_SHORT).withName("i16"),
+        MemoryLayout.sequenceLayout(4, JAVA_BYTE).withName("u8"),
+        MemoryLayout.sequenceLayout(4, JAVA_BYTE).withName("i8")
+    ).withName("_hb_var_int_t");
+
+    /*
+     * struct hb_glyph_position_t {
+     *     hb_position_t x_advance;
+     *     hb_position_t y_advance;
+     *     hb_position_t x_offset;
+     *     hb_position_t y_offset;
+     *     hb_var_int_t var;
+     * };
+     */
+     private static final StructLayout PositionLayout = MemoryLayout.structLayout(
+        JAVA_INT.withName("x_advance"),
+        JAVA_INT.withName("y_advance"),
+        JAVA_INT.withName("x_offset"),
+        JAVA_INT.withName("y_offset"),
+        VarIntLayout.withName("var")
+     ).withName("hb_glyph_position_t");
+
+    /**
+     * struct hb_glyph_info_t {
+     *     hb_codepoint_t codepoint;
+     *     hb_mask_t mask;
+     *     uint32_t cluster;
+     *     hb_var_int_t var1;
+     *     hb_var_int_t var2;
+     * };
+     */
+    private static final StructLayout GlyphInfoLayout = MemoryLayout.structLayout(
+        JAVA_INT.withName("codepoint"),
+        JAVA_INT.withName("mask"),
+        JAVA_INT.withName("cluster"),
+        VarIntLayout.withName("var1"),
+        VarIntLayout.withName("var2")
+    ).withName("hb_glyph_info_t");
+
+    private static VarHandle getVarHandle(StructLayout struct, String name) {
+        VarHandle h = struct.varHandle(PathElement.groupElement(name));
```

Comment:
Note that, strictly speaking, these combiners are _not_ required. If you just call `MemoryLayout::varHandle`, you get back a var handle that takes a `MemorySegment` and a `long` offset. So, you can, if you want, adapt the var handle so that the offset parameter becomes something else. But you could also just leave the var handle as is. Then, in the loop that is doing the access, you can do this:

```
for (int i = 0 ; i < limit ; i++) {
   x_offsetHandle.get(segment, PositionLayout.scale(0, i));
   y_offsetHandle.get(segment, PositionLayout.scale(0, i));
   ...
}
```
That is, use the offset hole to your advantage, to pass the struct base offset (obtained by scaling the enclosing struct layout by the value of the loop induction variable).

(That said, I understand that working with logical indices is a common operation, and that this is made a bit harder by the recent API changes. We should consider, as @JornVernee mentioned, adding back a more general `MemoryLayout::arrayElementVarHandle` which will give you the var handle you need - with coordinates `MemorySegment` and `long` - a logical index, not an offset).

Reply:
A PR which adds `MemoryLayout::arrayElementVarHandle` can be found here:
https://github.com/openjdk/jdk/pull/16272

With this, you can call the new method in order to create the var handle. The returned var handle will accept *two* long coordinate - the first is a base offset (as usual), the second in a logical index (what you need). The PR also adds plenty of narrative text describing how access to variable-length arrays should be performed using layouts (and also shows cases where the offset parameter is used in a non-trivial fashion).